src/number_theory/well_approximable.lean

```diff
@@ -0,0 +1,313 @@
+  simp only [nsmul_eq_mul],
+end
+
+@[to_additive]
+lemma smul_subset_of_coprime (han : (order_of a).coprime n) :
+  a • approx_order_of A n δ ⊆ approx_order_of A ((order_of a) * n) δ :=
+begin
+  simp_rw [approx_order_of, thickening_eq_bUnion_ball, ← image_smul, image_Union₂,
+    image_smul, smul_ball'', smul_eq_mul, mem_set_of_eq],
+  refine Union₂_subset_iff.mpr (λ b hb c hc, _),
+  simp only [mem_Union, exists_prop],
+  refine ⟨a * b, _, hc⟩,
+  rw ← hb at ⊢ han,
+  exact (commute.all a b).order_of_mul_eq_mul_order_of_of_coprime han,
+end
+
+@[to_additive vadd_eq_of_mul_dvd]
+lemma smul_eq_of_mul_dvd (hn : 0 < n) (han : (order_of a) * (order_of a) ∣ n) :
+  a • approx_order_of A n δ = approx_order_of A n δ :=
+begin
+  simp_rw [approx_order_of, thickening_eq_bUnion_ball, ← image_smul, image_Union₂,
+    image_smul, smul_ball'', smul_eq_mul, mem_set_of_eq],
+  replace han : ∀ {b : A}, order_of b = n → order_of (a * b) = n,
+  { intros b hb,
+    rw ← hb at han hn,
+    rwa [(commute.all a b).order_of_mul_eq_right_of_forall_prime_mul_dvd (order_of_pos_iff.mp hn)
+      (λ p hp hp', dvd_trans (mul_dvd_mul_right hp' $ order_of a) han)], },
+  let f : {b : A | order_of b = n} → {b : A | order_of b = n} := λ b, ⟨a * b, han b.property⟩,
+  have hf : surjective f,
+  { rintros ⟨b, hb⟩,
+    refine ⟨⟨a⁻¹ * b, _⟩, _⟩,
+    { rw [mem_set_of_eq, ← order_of_inv, mul_inv_rev, inv_inv, mul_comm],
+      apply han,
+      simpa, },
+    { simp only [subtype.mk_eq_mk, subtype.coe_mk, mul_inv_cancel_left], }, },
+  simpa only [f, mem_set_of_eq, subtype.coe_mk, Union_coe_set] using
+    hf.Union_comp (λ b, ball (b : A) δ),
+end
+
+end approx_order_of
+
+namespace unit_add_circle
+
+lemma mem_approx_add_order_of_iff {δ : ℝ} {x : unit_add_circle} {n : ℕ} (hn : 0 < n) :
+  x ∈ approx_add_order_of unit_add_circle n δ ↔
+  ∃ m < n, gcd m n = 1 ∧ ‖x - ↑((m : ℝ) / n)‖ < δ :=
+begin
+  haveI : fact ((0 : ℝ) < 1) := ⟨zero_lt_one⟩,
+  simp only [mem_approx_add_order_of_iff, mem_set_of_eq, ball, exists_prop, dist_eq_norm,
+    add_circle.add_order_of_eq_pos_iff hn, mul_one],
+  split,
+  { rintros ⟨y, ⟨m, hm₁, hm₂, rfl⟩, hx⟩, exact ⟨m, hm₁, hm₂, hx⟩, },
+  { rintros ⟨m, hm₁, hm₂, hx⟩, exact ⟨↑((m : ℝ) / n), ⟨m, hm₁, hm₂, rfl⟩, hx⟩, },
+end
+
+lemma mem_add_well_approximable_iff (δ : ℕ → ℝ) (x : unit_add_circle) :
+  x ∈ add_well_approximable unit_add_circle δ ↔
+  {n : ℕ | ∃ m < n, gcd m n = 1 ∧ ‖x - ↑((m : ℝ) / n)‖ < δ n}.infinite :=
+begin
+  simp only [mem_add_well_approximable_iff, ← nat.cofinite_eq_at_top, cofinite.blimsup_set_eq,
+    mem_set_of_eq],
+  refine iff_of_eq (congr_arg set.infinite $ ext (λ n, ⟨λ hn, _, λ hn, _⟩)),
+  { exact (mem_approx_add_order_of_iff hn.1).mp hn.2, },
+  { have h : 0 < n := by { obtain ⟨m, hm₁, hm₂, hm₃⟩ := hn, exact pos_of_gt hm₁, },
+    exact ⟨h, (mem_approx_add_order_of_iff h).mpr hn⟩, },
+end
+
+end unit_add_circle
+
+namespace add_circle
+
+variables {T : ℝ} [hT : fact (0 < T)]
+include hT
+
+local notation a `∤` b := ¬ a ∣ b
```

Comment:
Why isn't this a global notation? Or at least a global localized notation?

Reply:
Clearly it should be but I think this is best left alone till after the port since the notation `∣` (associated with the `has_dvd` typeclass) is in core.